frontend/catalyst/ag_primitives.py

```diff
@@ -318,7 +327,87 @@ def for_stmt(
     # Sometimes we unpack the results of nested tracing scopes so that the user doesn't have to
     # manipulate tuples when they don't expect it. Ensure set_state receives a tuple regardless.
     if not isinstance(results, tuple):
-        results = (results,)
+        # TODO: Do we have a use-case for this brnach?
+        results = (results,)  # pragma: nocover
+    set_state(results)
+
+
+def _call_catalyst_while(loop_test, loop_body, get_state, set_state, _nonlocals, _symbol_names):
+    """Dispatch to a Catalyst implementation of while loops."""
+
+    if _emulate_fallback_errors:
+        raise AutoGraphError("Emulated autograph fallback errror")
+
+    def _test(state):
+        old = get_state()
+        set_state(state)
+        res = loop_test()
+        set_state(old)
+        return res
+
+    @catalyst.while_loop(_test)
+    def _functional_while(iter_args):
+        set_state(iter_args)
+        loop_body()
+        return get_state()
+
+    iter_inits = get_state()
+    iter_results = _functional_while(iter_inits)
+    return iter_results
+
+
+def _call_python_while(loop_test, loop_body, get_state, _set_state, _nonlocals, _symbol_names):
+    """Fallback to a Python implementation of while loops."""
+
+    while loop_test():
+        loop_body()
+
+    return get_state()
+
+
+def while_stmt(loop_test, loop_body, get_state, set_state, nonlocals, symbol_names):
+    """An implementation of the AutoGraph 'while ..' statement. The interface is defined by
+    AutoGraph, here we merely provide an implementation of it in terms of Catalyst primitives."""
+
+    fallback = False
+
+    try:
+        results = _call_catalyst_while(
+            loop_test, loop_body, get_state, set_state, nonlocals, symbol_names
+        )
+
+    except Exception as e:  # pylint: disable=broad-exception-caught
+        if catalyst.autograph_strict_conversion:
+            raise e
+
+        # pylint: disable=import-outside-toplevel
+        import inspect
+        import textwrap
+
+        fallback = True
+
+        while_loop_info = get_source_code_info(inspect.stack()[1])
+
+        if not catalyst.autograph_ignore_fallbacks:
+            warnings.warn(
+                f"Tracing of an AutoGraph converted while loop failed with an exception:\n"
+                f"  {type(e).__name__}:{textwrap.indent(str(e), '    ')}\n"
+                f"\n"
+                f"The error ocurred within the body of the following for loop statement:\n"
+                f"{while_loop_info}"
+                f"\n"
+                f"To understand different types of JAX tracing errors, please refer to the "
+                f"guide at: https://jax.readthedocs.io/en/latest/errors.html\n"
+                f"\n"
+                f"If you did not intend for the conversion to happen, you may safely ignore "
+                f"this warning."
+            )
+
+    if fallback:
+        results = _call_python_while(
```

Comment:
I think it's safer to reset the state after a failed tracing attempt before falling back to Python.

Reply:
Indeed, done